src/main/scala/mesosphere/mesos/TaskBuilder.scala

```diff
@@ -208,41 +208,37 @@ class TaskBuilder(
   }
 
   protected def computeContainerInfo(hostPorts: Seq[Option[Int]]): Option[ContainerInfo] = {
-    if (runSpec.container.isEmpty && runSpec.ipAddress.isEmpty) None
-    else {
+    if (runSpec.container.isEmpty && runSpec.ipAddress.isEmpty) {
+      None
+    } else {
       val builder = ContainerInfo.newBuilder
 
       // Fill in Docker container details if necessary
       runSpec.container.foreach { c =>
-        val portMappings = c.docker.map { d =>
-          d.portMappings.map { pms =>
-            pms.zip(hostPorts).collect {
-              case (mapping, Some(hport)) =>
-                // Use case: containerPort = 0 and hostPort = 0
-                //
-                // For apps that have their own service registry and require p2p communication,
-                // they will need to advertise
-                // the externally visible ports that their components come up on.
-                // Since they generally know there container port and advertise that, this is
-                // fixed most easily if the container port is the same as the externally visible host
-                // port.
-                if (mapping.containerPort == 0) {
-                  mapping.copy(hostPort = Some(hport), containerPort = hport)
-                } else {
-                  mapping.copy(hostPort = Some(hport))
-                }
-            }
-          }
-        }
-
-        val containerWithPortMappings = portMappings match {
-          case None => c
-          case Some(newMappings) => c.copy(
-            docker = c.docker.map {
-              _.copy(portMappings = newMappings)
+        val containerWithPortMappings = c match {
+          case docker: Container.Docker => docker.copy(
```

Comment:
We need a container that zips its `portMappings` with the host's ports but otherwise not mutating the container. Is there a nicer way to do this in Scala? IMO copying is the best way to do here, but as we cannot copy the trait, we need to match the type first.